diff --git a/examples/pytorch/token-classification/run_ner.py b/examples/pytorch/token-classification/run_ner.py
index 65c26cd..177c6a4 100644
--- a/examples/pytorch/token-classification/run_ner.py
+++ b/examples/pytorch/token-classification/run_ner.py
@@ -29,6 +29,8 @@ import datasets
 import numpy as np
 from datasets import ClassLabel, load_dataset, load_metric

+from optimum.intel.nncf import NNCFAutoConfig
+
 import transformers
 from transformers import (
     AutoConfig,
@@ -469,6 +471,7 @@ def main():
             }

     # Initialize our Trainer
+    nncf_config = NNCFAutoConfig.from_json(training_args.nncf_config)
     trainer = Trainer(
         model=model,
         args=training_args,
@@ -477,6 +480,7 @@ def main():
         tokenizer=tokenizer,
         data_collator=data_collator,
         compute_metrics=compute_metrics,
+        nncf_config=nncf_config,
     )

     # Training
