diff --git a/modeling_utils.py b/modeling_utils.py
index 87cda0f..c7519da 100644
--- a/modeling_utils.py
+++ b/modeling_utils.py
@@ -1707,6 +1707,8 @@ PreTrainedModel.push_to_hub.__doc__ = PreTrainedModel.push_to_hub.__doc__.format
 )


+import nncf
+@nncf.torch.register_module()
 class Conv1D(nn.Module):
     """
     1D-convolutional layer as defined by Radford et al. for OpenAI GPT (and also used in GPT-2).
