diff --git a/trainer_callback.py b/trainer_callback.py
index e205c87..e66c654 100644
--- a/trainer_callback.py
+++ b/trainer_callback.py
@@ -468,6 +468,8 @@ class ProgressCallback(TrainerCallback):
     def on_step_end(self, args, state, control, **kwargs):
         if state.is_local_process_zero:
             self.training_bar.update(state.global_step - self.current_step)
+            if hasattr(state, "curr_loss"):
+                self.training_bar.set_postfix(loss=state.curr_loss)
             self.current_step = state.global_step

     def on_prediction_step(self, args, state, control, eval_dataloader=None, **kwargs):
